File: convert/C++/valute.cpp
#include <iostream>
#include <string>
#include <vector>
#include <curl/curl.h>
#include "json.hpp" // Скачайте с https://github.com/nlohmann/json

using json = nlohmann::json;
using namespace std;

// Функция для обработки ответа от сервера
size_t WriteCallback(void* contents, size_t size, size_t nmemb, void* userp) {
    ((string*)userp)->append((char*)contents, size * nmemb);
    return size * nmemb;
}

// Универсальная функция для выполнения GET-запроса
string performRequest(string url) {
    CURL* curl;
    CURLcode res;
    string readBuffer;

    curl = curl_easy_init();
    if (curl) {
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);
        res = curl_easy_perform(curl);
        curl_easy_cleanup(curl);
    }
    return readBuffer;
}

int main() {
    string apiKey = "cd3b386195c6857e09dbfde5";
    
    // 1. Получаем список всех кодов валют
    string codesUrl = "https://v6.exchangerate-api.com/v6/" + apiKey + "/codes";
    string codesResponse = performRequest(codesUrl);
    auto codesJson = json::parse(codesResponse);

    if (codesJson["result"] != "success") {
        cout << "Ошибка API при загрузке кодов!" << endl;
        return 1;
    }

    auto supportedCodes = codesJson["supported_codes"];

    auto getValidCurrency = [&](string prompt) {
        while (true) {
            cout << prompt;
            string input;
            cin >> input;
            for (auto& c : input) c = toupper(c);

            // Проверка на валидность
            for (auto& item : supportedCodes) {
                if (item[0] == input) return input;
            }

            // Если не найдено — поиск подсказок
            cout << "❌ Валюта '" << input << "' не найдена." << endl;
            string prefix = input.substr(0, 2);
            cout << "Возможно, вы имели в виду:" << endl;
            bool found = false;
            for (auto& item : supportedCodes) {
                string code = item[0];
                if (code.substr(0, 2) == prefix) {
                    cout << "  • " << code << " (" << item[1] << ")" << endl;
                    found = true;
                }
            }
            if (!found) cout << "Вариантов не найдено." << endl;
            cout << "--------------------" << endl;
        }
    };

    // 2. Логика конвертации
    string base = getValidCurrency("Введите исходную валюту (USD): ");
    string target = getValidCurrency("Введите целеную валюту (RUB): ");
    
    double amount;
    cout << "Введите сумму: ";
    cin >> amount;

    string pairUrl = "https://v6.exchangerate-api.com/v6/" + apiKey + "/pair/" + base + "/" + target;
    string pairResponse = performRequest(pairUrl);
    auto pairJson = json::parse(pairResponse);

    if (pairJson["result"] == "success") {
        double rate = pairJson["conversion_rate"];
        cout << "\n✅ Итог: " << amount << " " << base << " = " 
             << (amount * rate) << " " << target << " (Курс: " << rate << ")" << endl;
    }

    return 0;
}

File: convert/python/valute.py
import requests

def get_supported_codes(api_key):
    url = f"https://v6.exchangerate-api.com/v6/{api_key}/codes"
    try:
        response = requests.get(url)
        data = response.json()
        if data['result'] == 'success':
            return data['supported_codes']
    except:
        return []
    return []
def find_suggestions(user_input, codes_list):
    prefix = user_input[:2].upper()
    suggestions = [f"{code} ({name})" for code, name in codes_list if code.startswith(prefix)]
    return suggestions
def convert_currency():
    api_key = "cd3b386195c6857e09dbfde5" 
    all_codes_data = get_supported_codes(api_key)
    valid_codes = [item[0] for item in all_codes_data]
    print("конвертер валют ")
    def get_valid_currency(prompt):
        while True:
            code = input(prompt).upper()
            if code in valid_codes:
                return code
            else:
                print(f"Валюта '{code}' не найдена")
                suggestions = find_suggestions(code, all_codes_data)
                if suggestions:
                    print("Возможно, вы имели в виду:")
                    for s in suggestions:
                        print(f"  • {s}")
                else:
                    print("подходящих вариантов не найдено")
                print("-" * 20)

    base = get_valid_currency("Введите исходную валюту (напр. USD): ")
    target = get_valid_currency("Введите целевую валюту (напр. RUB): ")
    try:
        amount = float(input(f"Сумма в {base}: "))
    except ValueError:
        print("Ошибка: введите число.")
        return
    url = f"https://v6.exchangerate-api.com/v6/{api_key}/pair/{base}/{target}"
    response = requests.get(url).json()
    if response['result'] == 'success':
        rate = response['conversion_rate']
        result = amount * rate
        print(f"\n✅ Результат: {amount} {base} = {result:.2f} {target}")
    else:
        print("Произошла ошибка при получении курса")
if __name__ == "__main__":
    convert_currency()
